src/move/emojicoin_arena/sources/emojicoin_arena.move

```diff
@@ -0,0 +1,852 @@
+    const DEFAULT_AVAILABLE_REWARDS: u64 = 1000 * 100_000_000;
+    const DEFAULT_MAX_MATCH_PERCENTAGE: u64 = 50;
+    const DEFAULT_MAX_MATCH_AMOUNT: u64 = 5 * 100_000_000;
+
+    struct Melee has store {
+        /// 1-indexed for conformity with emojicoin market ID indexing.
+        melee_id: u64,
+        /// Metadata for market with lower market ID comes first.
+        market_metadatas: vector<MarketMetadata>,
+        /// In microseconds.
+        start_time: u64,
+        /// How long melee lasts after start time.
+        duration: u64,
+        /// Max percentage of user's input amount to match in octas, when locking in.
+        max_match_percentage: u64,
+        /// Maximum amount of APT to match in octas, when locking in.
+        max_match_amount: u64,
+        /// Amount of rewards that are available to claim for this melee while it is still active.
+        /// Measured in octas, conditional on vault balance.
+        available_rewards: u64,
+        /// All entrants who have entered the melee, used as a set.
+        all_entrants: SmartTable<address, Nil>,
+        /// Active entrants in the melee, used as a set.
+        active_entrants: SmartTable<address, Nil>,
+        /// Entrants who have exited the melee, used as a set.
+        exited_entrants: SmartTable<address, Nil>,
+        /// Entrants who have locked in, used as a set. If user exits before melee ends they are
+        /// removed from this set. If they exit after the melee ends, they are not removed.
+        locked_in_entrants: SmartTable<address, Nil>,
+        /// Number of melee-specific swaps.
+        n_melee_swaps: Aggregator<u64>,
+        /// Volume of melee-specific swaps in octas.
+        melee_swaps_volume: Aggregator<u128>,
+        /// Amount of emojicoin 0 locked in all melee escrows for the melee.
+        emojicoin_0_locked: Aggregator<u64>,
+        /// Amount of emojicoin 1 locked in all melee escrows for the melee.
+        emojicoin_1_locked: Aggregator<u64>
+    }
+
+    struct MeleeEscrow<phantom Coin0, phantom LP0, phantom Coin1, phantom LP1> has key {
+        /// Corresponding `Melee.melee_id`.
+        melee_id: u64,
+        /// Emojicoin 0 holdings.
+        emojicoin_0: Coin<Coin0>,
+        /// Emojicoin 1 holdings.
+        emojicoin_1: Coin<Coin1>,
+        /// Volume of user's melee-specific swaps in octas.
+        melee_swaps_volume: u128,
+        /// Number of swaps user has executed during the melee.
+        n_melee_swaps: u64,
+        /// Cumulative amount of APT entered into the melee since the most recent deposit into an
+        /// empty escrow. Inclusive of total amount matched from locking in since most recent
+        /// deposit into an empty escrow. Reset to 0 upon exit.
+        octas_entered: u64,
+        /// Cumulative amount of APT matched since most recent deposit into an empty escrow, reset
+        /// to 0 upon exit. Must be paid back in full when tapping out.
+        octas_matched: u64
```

Comment:
@CRBl69 @xbtmatt I've updated these fields since the last review: does this work?